### DISCUSSION TITLE
Gradient Caption Magick.NET

### DISCUSSION THREAD

**Discussion by Nihaarn1997** [dlemstra/Magick.NET#781] [Help]
Hi,
I am working on a C# project which currently invokes IM commands for creating images with titles. I'm now switching over to using Magick.NET.
Having trouble converting the following command to Magick.NET - 
convert -size 374x550 -background none -font "PathToFontFile" -gravity center caption:"Title" ( -clone 0 -alpha opaque -sparse-color Barycentric "0, 0 #070504  %[fx:w-1],0 #000000" ) ( -clone 0 -alpha transparent ) -reverse -compose dstover -composite "PathToLocation"

I'd really appreciate any help as I have been stuck on this for weeks now.

Thanks
Nihaar Nandedkar

(Note: Below is an example of what I am trying to achieve)
![hEDTOverlay#cdcd37#37cdffRubik-Medium](https://user-images.githubusercontent.com/74175275/98669756-c221af80-2377-11eb-9acd-80021ceaac2f.png)

**Comment by dlemstra** [comment]
What have you tried so far? At what point are you stuck?

**Comment by Nihaarn1997** [comment]
What we have tried:
1. Successfully overlaid gradient text over an image using the IM command as stated in the issue
2. Overlay single color text over an image using the read() and composite() methods of Magick.NET


Where we are stuck:
1. Switching over from using IM commands to Magick.NET library
2. Below is an example of how we achieved vertical, horizontal and arc gradient overlay text

convert -size 374x550 -background none -font "PathToFontFile" -gravity center caption:"Nations Cup" ( -clone 0 -alpha opaque -sparse-color Barycentric **"0, 0 #070504  %[fx:w-1],0 #000000"** ) ( -clone 0 -alpha transparent ) -reverse -compose dstover -composite "PathToLocation"

convert -size 374x550 -background none -font "PathToFontFile" -gravity center caption:"Nations Cup" ( -clone 0 -alpha opaque -sparse-color Barycentric **"0, 0 #070504  0,%[fx:h-1], #000000"** ) ( -clone 0 -alpha transparent ) -reverse -compose dstover -composite "PathToLocation"

convert -size 374x550 -background none -font "PathToFontFile" -gravity center caption:"Nations Cup" ( +clone -sparse-color Barycentric **"0, %h #070504  %w,0 #000000"** ) -compose In -composite -virtual-pixel transparent -distort arc 80 ( +clone -background none -shadow 100x2+4+4 ) +swap -background none -compose over -layers merge +repage "PathToLocation"

3. The gradient effect (vertical, horizontal & arc) requires 2 colors as input and a few other parameters along with it, not sure how to pass to to the below code

                MagickReadSettings settings = new MagickReadSettings();
                settings.FillColor = MagickColor.FromRgb(r, g, b);
                settings.Font = ConfigurationManager.AppSettings["fontpath"] + singleFont.SystemFontName;
                settings.Width = writableDetails.WidthWritable;
                settings.Height = writableDetails.HeightWritable;
                settings.BackgroundColor = MagickColor.FromRgba(255, 255, 255, 0);
                settings.TextGravity = Gravity.Center;
                settings.TextInterlineSpacing = writableDetails.InterlineSpace;

**Comment by dlemstra** [comment]
The example command that you have does not create the sample image that you provided? Here is a translation of your first command:

```
convert -size 374x550 -background none -font "PathToFontFile" -gravity center caption:"Nations Cup"
( -clone 0 -alpha opaque -sparse-color Barycentric "0, 0 #070504 %[fx:w-1],0 #000000" )
( -clone 0 -alpha transparent ) -reverse -compose dstover -composite "PathToLocation"
```

```C#
var readSettings = new MagickReadSettings
{
    // -size 374x550
    Width = 374,
    Height = 550,

    // -background none
    BackgroundColor = MagickColors.None,

    // -font "PathToFontFile"
    Font = "PathToFontFile",

    // -gravity center
    TextGravity = Gravity.Center,
};

// caption:"Nations Cup"
using (var image = new MagickImage("caption:Nations Cup", readSettings))
{
    // -clone 0
    using (var clone1 = image.Clone())
    {
        // -alpha opaque
        clone1.Alpha(AlphaOption.Opaque);

        // -sparse-color Barycentric "0, 0 #070504 %[fx:w-1],0 #000000"
        var sparseColorArgs = new[]
        {
            // 0, 0 #070504
            new SparseColorArg(0, 0, new MagickColor("#070504")),

            // %[fx:w-1],0 #000000
            new SparseColorArg(image.Width - 1, 0, new MagickColor("#000000")),
        };

        // -sparse-color Barycentric
        clone1.SparseColor(SparseColorMethod.Barycentric, sparseColorArgs);

        // -clone 0
        using (var clone2 = image.Clone())
        {
            // -alpha transparent
            clone2.Alpha(AlphaOption.Transparent);

            // -reverse changes the order of the images. The order was:
            // image, clone1, clone2 but this is not clone2, clone1, image.

            // -compose dstover -composite
            // Because there are there images some special magick is happening.
            using (var clone3 = clone2.Clone())
            {
                clone2.Composite(clone1, CompositeOperator.DstOver);

                clone2.Composite(image, CompositeOperator.CopyAlpha);

                clone3.Composite(clone2, CompositeOperator.Over);

                // "PathToLocation"
                clone3.Write("PathToLocation");
            }
        }
    }
}
```

Feel free to ask if you have any further questions.

**Comment by Nihaarn1997** [comment]
Thanks for the help, the above code worked perfectly for both horizontal and vertical gradient effects.
Replaced 
```
SparseColorArg[] sparseColorArgs = new SparseColorArg[2];
sparseColorArgs[0] = new SparseColorArg(0, 0, new MagickColor(singleColor1.TextColor));
if (gradient_vertical)
                sparseColorArgs[1] = new SparseColorArg(0, imgMain.Height - 1, new MagickColor(singleColor2.TextColor));
else if (gradient_horizontal)
                sparseColorArgs[1] = new SparseColorArg(imgMain.Width - 1, 0, new MagickColor(singleColor2.TextColor));
```
We need help with overlaying the caption in an arc where we pass the angle of arc as parameter to the IM command. Below is the command for it

convert -size 374x550 -background none -font "PathToFontFile" -gravity center caption:"Nations Cup" ( +clone -sparse-color Barycentric "0, %h #070504 %w,0 #000000" ) -compose In -composite -virtual-pixel transparent -distort arc 80 ( +clone -background none -shadow 100x2+4+4 ) +swap -background none -compose over -layers merge +repage "PathToLocation"

Can you please help us in translating this command to Magick.NET code?

**Comment by dlemstra** [comment]
What have you tried yourself and where are you stuck?

**Reply by Nihaarn1997** [reply]
Sorry for replying so late. Had a long weekend. Below is a code snippet of what we have tried - 

                MagickReadSettings settings = new MagickReadSettings();
                settings.BackgroundColor = MagickColors.None;
                settings.FillColor = new MagickColor("#7fffd4");
                settings.Font = ConfigurationManager.AppSettings["fontpath"] + singleFont.SystemFontName;
                settings.Width = writableDetails.WidthWritable;
                settings.Height = writableDetails.HeightWritable;
                settings.TextGravity = Gravity.Center;
                if (writableDetails.InterlineSpace != 0)
                {
                    settings.TextInterlineSpacing = writableDetails.InterlineSpace;
                }

                if (singleFont.StrokeWidth != 0)
                {
                    settings.StrokeColor = MagickColors.Violet;
                    settings.StrokeWidth = singleFont.StrokeWidth;
                }

                MagickImage background = new MagickImage(imageDetails.ImagePath);

                MagickImage imgMain = new MagickImage(MagickColors.None, writableDetails.WidthWritable, writableDetails.HeightWritable);
                imgMain.ColorFuzz = new Percentage(100);
                imgMain.Transparent(MagickColors.White);
                imgMain.Read("caption:" + "Bledisloe Cup", settings);
                imgMain.Composite(imgMain, Gravity.Center, CompositeOperator.Over);

                MagickImage clone1 = (MagickImage)imgMain.Clone();
                    clone1.Alpha(AlphaOption.Opaque);
                    SparseColorArg[] sparseColorArgs = new SparseColorArg[2];
                    if (arc)
                    {
                        sparseColorArgs[0] = new SparseColorArg(0, imgMain.Height, new MagickColor("#7fffd4"));
                        sparseColorArgs[1] = new SparseColorArg(imgMain.Width, 0, new MagickColor("#00FF00"));
                        clone1.SparseColor(SparseColorMethod.Barycentric, sparseColorArgs);
                        MagickImage clone2 = (MagickImage)imgMain.Clone();
                        clone2.VirtualPixelMethod = VirtualPixelMethod.Transparent;
                        clone2.Distort(DistortMethod.Arc, 80);

                        MagickImage imgShadow = (MagickImage)imgMain.Clone();
                        imgShadow.Shadow(5, 5, 3, (Percentage)80);
                            imgShadow.BackgroundColor = MagickColors.None;
                            imgShadow.Composite(imgMain, -1, -1, CompositeOperator.Over);

                        imgShadow.Composite(clone2, CompositeOperator.MinusDst);

                        //MagickImage clone3 = (MagickImage)imgShadow.Clone();
                        //imgShadow.Composite(clone1, CompositeOperator.DstOver);
                        //imgShadow.Composite(imgMain, CompositeOperator.CopyAlpha);
                        //clone3.Composite(imgShadow, CompositeOperator.Over);
                        background.Composite(imgShadow, Gravity.Center, CompositeOperator.Over);
                        background.Write((targetpath + "7fffd4" + "00FF00" + singleFont.SystemFontName.Replace(".ttf", "") + imageDetails.ImageCode + ".jpg").Replace("#", ""));
                    }
                    else
                    {
                        
                        sparseColorArgs[0] = new SparseColorArg(0, 0, new MagickColor("#7fffd4"));
                        if (gradient_horizontal)
                            sparseColorArgs[1] = new SparseColorArg(imgMain.Width - 1, 0, new MagickColor("#00FF00"));
                        else if (gradient_vertical)
                            sparseColorArgs[1] = new SparseColorArg(0, imgMain.Height - 1, new MagickColor("#00FF00"));


                        clone1.SparseColor(SparseColorMethod.Barycentric, sparseColorArgs);

                        MagickImage clone2 = (MagickImage)imgMain.Clone();
                        clone2.Alpha(AlphaOption.Transparent);

                        MagickImage clone3 = (MagickImage)clone2.Clone();
                        clone2.Composite(clone1, CompositeOperator.DstOver);
                        clone2.Composite(imgMain, CompositeOperator.CopyAlpha);
                        clone3.Composite(clone2, CompositeOperator.Over);
                        background.Composite(clone3, Gravity.Center, CompositeOperator.Over);
                        background.Write((targetpath + "7fffd4" + "00FF00" + singleFont.SystemFontName.Replace(".ttf", "") + imageDetails.ImageCode + ".jpg").Replace("#", ""));
                    }
                    objectadd.localurl = (targetpath.Replace(ConfigurationManager.AppSettings["targetpath"], "http://localhost/ImageRepo/").Replace("\\", "/") + "#7fffd4" + "#00FF00" + singleFont.SystemFontName.Replace(".ttf", "") + imageDetails.ImageCode + ".jpg").Replace("#", "");

                }

**The else part of the above code is working absolutely fine as suggested by you earlier**
We are trying to achieve the effect of having the text get overlaid in an arc but we are stuck. The 2 images below show what we have achieved vs what we want to achieve.

Output of using Magick.NET
![7fffd400FF00Futura-Bold-Condensed-ObliqueH8](https://user-images.githubusercontent.com/74175275/101344548-a55f9580-38ab-11eb-978e-b435241fac17.jpg)

Output of using ImageMagick Command
![524c54370537HelveticaNeueLTStd-MdCnE1A](https://user-images.githubusercontent.com/74175275/101344552-a7c1ef80-38ab-11eb-8e2b-bdfe4f72881b.jpg)

**Reply by dlemstra** [reply]
Where is your input image and what is the ImageMagick command that you executed for this last image?

**Reply by Nihaarn1997** [reply]
There is no input image as such. The system we have developed will first create a .png file with the overlaying text using the Convert Command of ImageMagick. The system will then pick up a random image from the repository and overlay the generated png text file on top of it using the Composite command.

The ImageMagick command is similar to the one we have provided earlier (November 11). The system just changes the parameters of that command based on which image has been selected from the repository and the length of the text that needs to be overlaid. Below is an example of the ImageMagick command being used.

convert -size widthxheight -background none -font "PathToFontFile" -gravity center caption:"Bledisloe Cup" ( +clone -sparse-color Barycentric "0, %h #070504 %w,0 #000000" ) -compose In -composite -virtual-pixel transparent -distort arc 80 ( +clone -background none -shadow 100x2+4+4 ) +swap -background none -compose over -layers merge +repage "PathToLocation"

We are trying to convert this command to Magick.NET code and that's where we are stuck. From your previous answers we were able to achieve the gradient color effect for the text but we aren't able to get that same text in an arc (angle specified in the command). We tried using the Distort method but as you can see, the output was not satisfactory.

**Reply by dlemstra** [reply]
I don't understand your issue. Please provide me with the following this:
- input image
- imagemagick command
- output of the imagemagick command
- magick.net code
- output of the magick.net code

**Reply by Nihaarn1997** [reply]
Hi please find an example of how we use the ImageMagick commands to generate the desired output.

Example contains overlay image, background image, final image and the command which is shown below

convert -background none -fill #f9f9fa -font "C:\Users\gnmumhelp\Downloads\SysFonts\LeagueGothic-Regular.ttf" -size 1536x864 -gravity center caption:"Bledisloe Cup" ( +clone -sparse-color Barycentric "0, %h #f9f9fa  %w,0 #cdff37" ) -compose In -composite -virtual-pixel transparent -distort arc 50 ( +clone -background none -shadow 100x2+4+4 ) +swap -background none -compose over -layers merge +repage "C:\Users\gnmumhelp\Downloads\ImageMagick\Overlay\ABledisloe CupOverlayH8#f9f9fa#cdff37LeagueGothic-Regular.png"

composite -gravity center "C:\Users\gnmumhelp\Downloads\ImageMagick\Overlay\ABledisloe CupOverlayH8#f9f9fa#cdff37LeagueGothic-Regular.png" C:\Users\gnmumhelp\Downloads\BackgroundRepo\Sport\Rugby\Dark\2SportRugbyH8Dark.jpg "C:\Images\30000000549780901\gradient\f9f9facdff37LeagueGothic-RegularH8A.jpg"

Below is the Overlay Image
![ABledisloe CupOverlayH8#f9f9fa#cdff37LeagueGothic-Regular](https://user-images.githubusercontent.com/74175275/101617041-1f238a80-3a36-11eb-8147-e330a28515ca.png)

This is the background image
![2SportRugbyH8Dark](https://user-images.githubusercontent.com/74175275/101617055-25b20200-3a36-11eb-982c-d69d9bec3e51.jpg)

And this is the final output image
![f9f9facdff37LeagueGothic-RegularH8A](https://user-images.githubusercontent.com/74175275/101617080-31052d80-3a36-11eb-9fa4-29aa73bd8428.jpg)

For your reference, the below image describes the functionality of our system in brief
<img width="898" alt="SeeThis" src="https://user-images.githubusercontent.com/74175275/101617188-572acd80-3a36-11eb-9cf0-8767529ac3ec.png">


For the Magick.NET code and the output of the Magick.NET code, please refer below
            
            MagickReadSettings settings = new MagickReadSettings();
            settings.BackgroundColor = MagickColors.None;
            settings.FillColor = new MagickColor("#f9f9fa");
            settings.Font = "C:\\Users\\gnmumhelp\\Downloads\\SysFonts\\LeagueGothic-Regular.ttf";
            settings.Width = 1536;
            settings.Height = 864;
            settings.TextGravity = Gravity.Center;

            MagickImage background = new MagickImage("C:\\Users\\gnmumhelp\\Downloads\\BackgroundRepo\\Sport\\Rugby\\Dark\\2SportRugbyH8Dark.jpg");

            MagickImage imgMain = new MagickImage(MagickColors.None, 1536, 864);
            imgMain.ColorFuzz = new Percentage(100);
            imgMain.Transparent(MagickColors.White);
            imgMain.Read("caption:" + "Bledisloe Cup", settings);
            imgMain.Composite(imgMain, Gravity.Center, CompositeOperator.Over);

            MagickImage clone1 = (MagickImage)imgMain.Clone();
            clone1.Alpha(AlphaOption.Opaque);
            SparseColorArg[] sparseColorArgs = new SparseColorArg[2];
            sparseColorArgs[0] = new SparseColorArg(0, imgMain.Height, new MagickColor("#f9f9fa"));
            sparseColorArgs[1] = new SparseColorArg(imgMain.Width, 0, new MagickColor("#cdff37"));
            clone1.SparseColor(SparseColorMethod.Barycentric, sparseColorArgs);
            MagickImage clone2 = (MagickImage)imgMain.Clone();
            clone2.VirtualPixelMethod = VirtualPixelMethod.Transparent;
            clone2.Distort(DistortMethod.Arc, 50);

            MagickImage imgShadow = (MagickImage)imgMain.Clone();
            imgShadow.Shadow(5, 5, 3, (Percentage)50);
            imgShadow.BackgroundColor = MagickColors.None;
            imgShadow.Composite(imgMain, -1, -1, CompositeOperator.Over);

            imgShadow.Composite(clone2, CompositeOperator.MinusDst);

            //MagickImage clone3 = (MagickImage)imgShadow.Clone();
            //imgShadow.Composite(clone1, CompositeOperator.DstOver);
            //imgShadow.Composite(imgMain, CompositeOperator.CopyAlpha);
            //clone3.Composite(imgShadow, CompositeOperator.Over);
            background.Composite(imgShadow, Gravity.Center, CompositeOperator.Over);
            background.Write("C:\\Images\\30000000549780901\\gradient\\" + "f9f9fa" + "cdff37" + "LeagueGothic-Regular" + "H8A" + ".jpg");

And below is the output of the above code
![f9f9facdff37LeagueGothic-RegularH8A](https://user-images.githubusercontent.com/74175275/101619354-29935380-3a39-11eb-9c38-cf1b523d358f.jpg)

**Comment by dlemstra** [comment]
Your commands could be translated to this:

```C#
var settings = new MagickReadSettings
{
    BackgroundColor = MagickColors.None, // -background none
    FillColor = new MagickColor("#f9f9fa"), // -fill #f9f9fa
    Font = "LeagueGothic-Regular.otf", // -font "LeagueGothic-Regular.otf"
    Width = 1536,
    Height = 864, // -size 1536x864
    TextGravity = Gravity.Center, // -gravity center
};

using (var caption = new MagickImage("caption:Bledisloe Cup", settings))
{
    using (var sparseColorImage = caption.Clone()) // +clone
    {
        var sparseColorArgs = new SparseColorArg[2];
        sparseColorArgs[0] = new SparseColorArg(0, caption.Height, new MagickColor("#f9f9fa")); // 0, %h #f9f9fa
        sparseColorArgs[1] = new SparseColorArg(caption.Width, 0, new MagickColor("#cdff37")); // %w,0 #cdff37

        sparseColorImage.SparseColor(SparseColorMethod.Barycentric, sparseColorArgs); // -sparse - color Barycentric

        caption.Composite(sparseColorImage, CompositeOperator.In); // -compose In -composite
    }

    caption.VirtualPixelMethod = VirtualPixelMethod.Transparent; // -virtual-pixel
    caption.Distort(DistortMethod.Arc, 50); // -distort arc 50

    using (var overlay = caption.Clone()) // +clone
    {
        // not sure why this needs to be black instead of none.
        overlay.BackgroundColor = MagickColors.Black;
        overlay.Shadow(4, 4, 2, new Percentage(100)); // -shadow 100x2+4+4

        // Easier than -compose over -layers merge
        overlay.Composite(caption, CompositeOperator.Over);

        using (var background = new MagickImage("background.jpg"))
        {
            // composite -gravity center
            background.Composite(overlay, Gravity.Center, CompositeOperator.Over);
            background.Write("a.jpg");
        }
    }
}
```

**Reply by Nihaarn1997** [reply]
The above code worked perfectly and generated the output image as required.
Thanks a lot for all the help you've provided over the past couple of weeks.
Cheers!
